src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageLoaderMatcherPolicy.cs

```diff
@@ -0,0 +1,167 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.ActionConstraints;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.AspNetCore.Mvc.Routing;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.Matching;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure
+{
+    internal class PageLoaderMatcherPolicy : MatcherPolicy, IEndpointSelectorPolicy
+    {
+        private readonly IPageLoader _loader;
+        private readonly MvcEndpointInvokerFactory _invokerFactory;
+
+        public PageLoaderMatcherPolicy(IPageLoader loader, MvcEndpointInvokerFactory invokerFactory)
+        {
+            if (loader == null)
+            {
+                throw new ArgumentNullException(nameof(loader));
+            }
+
+            if (invokerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(invokerFactory));
+            }
+
+            _loader = loader;
+            _invokerFactory = invokerFactory;
+        }
+
+        public override int Order => int.MinValue + 100;
+
+        public bool AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
+        {
+            if (endpoints == null)
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+            
+            for (var i = 0; i < endpoints.Count; i++)
+            {
+                var page = endpoints[i].Metadata.GetMetadata<PageActionDescriptor>();
+                if (page != null)
+                {
+                    // Found a page
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public Task ApplyAsync(HttpContext httpContext, EndpointSelectorContext context, CandidateSet candidates)
+        {
+            if (httpContext == null)
+            {
+                throw new ArgumentNullException(nameof(httpContext));
+            }
+
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (candidates == null)
+            {
+                throw new ArgumentNullException(nameof(candidates));
+            }
+
+            for (var i = 0; i < candidates.Count; i++)
+            {
+                ref var candidate = ref candidates[i];
+                var endpoint = (RouteEndpoint)candidate.Endpoint;
+
+                var page = endpoint.Metadata.GetMetadata<PageActionDescriptor>();
+                if (page != null)
+                {
+                    var compiled = _loader.Load(page);
+
+                    // This is wrong - we don't want to allocate this per-request. #YOLO
+                    RequestDelegate requestDelegate = (c) =>
```

Comment:
Imagine that line 88-159 are hidden away somewhere else and you get the main idea.